compiler/psi/src/org/jetbrains/kotlin/parsing/KotlinParsing.java

```diff
@@ -2423,43 +2446,58 @@ boolean tryParseTypeArgumentList(TokenSet extraRecoverySet) {
                 advance(); // MUL
             }
             else {
-                parseTypeRef(extraRecoverySet);
+                parseTypeRef();
             }
             projection.done(TYPE_PROJECTION);
-            if (!at(COMMA)) break;
-            advance(); // COMMA
-            if (at(GT)) {
-                break;
+
+            if (at(COMMA)) {
+                advance(); // COMMA
+            }
+            else if (!at(GT) && !at(GTEQ)) {
+                // if it is not strict, then we want to fail fast as any error
+                // will roll back this type argument list anyway
+                if (!strict || atSet(TYPE_ARGUMENT_LIST_RECOVERY_SET) || eof()) break;
+
+                error("Expecting comma or '>'");
             }
+
+            if (at(GT) || at(GTEQ)) break;
         }
 
-        boolean atGT = at(GT);
-        if (!atGT) {
-            error("Expecting a '>'");
+        if (at(GT)) {
+            myBuilder.disableJoiningComplexTokens();
+            advance(); // GT
+            myBuilder.restoreJoiningComplexTokensState();
         }
         else {
-            advance(); // GT
+            error("Expecting a '>'");
         }
+
         myBuilder.restoreNewlinesState();
-        return atGT;
+
+        boolean ok = empty || strict || !myBuilder.hasErrorsAfter(errorScope);
+
+        errorScope.drop();
+
+        return ok;
     }
 
     /*
      * functionType
      *   : (type ".")? "(" parameter{","}? ")" "->" type?
      *   ;
      */
-    private void parseFunctionType(PsiBuilder.Marker functionType) {
-        parseFunctionTypeContents(functionType).done(FUNCTION_TYPE);
+    private void parseFunctionType(PsiBuilder.Marker functionType, TokenSet recoverySet) {
```

Comment:
It was somewhat unimportant in the original approach, but has now become more important in the updated approach. What it does, is returning to `tryParseTypeArgumentList` sooner without having consumed as many tokens as it would have. If the token is contained in the `TYPE_ARGUMENT_LIST_RECOVERY_SET`, it will abort parsing the type argument list and return `TypeArgumentListKind.NONE`.